cpp/src/arrow/util/thread-pool.h

```diff
@@ -0,0 +1,147 @@
+#include <memory>
+#include <mutex>
+#include <thread>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+#include "arrow/status.h"
+#include "arrow/util/macros.h"
+
+namespace arrow {
+
+// Set the number of worker threads used by the process-global thread pool
+// for CPU-bound tasks.
+ARROW_EXPORT Status SetCPUThreadPoolCapacity(size_t threads);
+
+namespace internal {
+
+namespace detail {
+
+// Needed because std::packaged_task is not copyable and hence not convertible
+// to std::function.
+template <typename R, typename... Args>
+struct packaged_task_wrapper {
+  using PackagedTask = std::packaged_task<R(Args...)>;
+
+  explicit packaged_task_wrapper(PackagedTask&& task)
+      : task_(std::make_shared<PackagedTask>(std::forward<PackagedTask>(task))) {}
+
+  void operator()(Args&&... args) { return (*task_)(std::forward<Args>(args)...); }
+  std::shared_ptr<PackagedTask> task_;
+};
+
+}  // namespace detail
+
+class ThreadPool {
+ public:
+  // Construct a thread pool with the given number of worker threads
+  static Status Make(size_t threads, std::shared_ptr<ThreadPool>* out);
+
+  // Destroy thread pool; the pool will first be shut down
+  ~ThreadPool();
+
+  // Dynamically change the number of worker threads.
+  // This function returns quickly, but it may take more time before the
+  // thread count is fully adjusted.
+  Status SetCapacity(size_t threads);
+
+  // Shutdown the pool.  Once the pool starts shutting down, new tasks
+  // cannot be submitted anymore.
+  // If "wait" is true, shutdown waits for all pending tasks to be finished.
+  // If "wait" is false, workers are stopped as soon as currently executing
+  // tasks are finished.
+  Status Shutdown(bool wait = true);
+
+  // Spawn a fire-and-forget task on one of the workers.
+  template <typename Function>
+  Status Spawn(Function&& func) {
+    return SpawnReal(std::forward<Function>(func));
+  }
+
+  // Submit a callable and arguments for execution.  Return a future that
+  // will return the callable's result value once.
+  // The callable's arguments are copied before execution.
+  // Since the function is variadic and needs to return a result (the future),
+  // an exception is raised if the task fails spawning (which currently
+  // only occurs if the ThreadPool is shutting down).
+  template <typename Function, typename... Args,
+            typename Result = typename std::result_of<Function && (Args && ...)>::type>
+  std::future<Result> Submit(Function&& func, Args&&... args) {
+    // Trying to templatize std::packaged_task with Function doesn't seem
+    // to work, so go through std::bind to simplify the packaged signature
+    using PackagedTask = std::packaged_task<Result()>;
+    auto task = PackagedTask(std::bind(std::forward<Function>(func), args...));
+    auto fut = task.get_future();
+
+    Status st = SpawnReal(detail::packaged_task_wrapper<Result>(std::move(task)));
+    if (!st.ok()) {
+      throw std::runtime_error(st.ToString());
```

Comment:
How could this exception manifest in real code, or would it be an esoteric failure (a la DCHECK)?

Reply:
It manifests if you call `Submit` after `Shutdown`. It's a programmer error, though, not a random runtime condition.

Reply:
Makes sense. If you get a chance, adding a code comment there to this effect would be great